src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/ImportOrganizationUserCommand.cs

```diff
@@ -0,0 +1,337 @@
+
+        await AddNewUsers(organization, newUsers, importUserData);
+
+        await ImportGroups(organization, groups, importUserData);
+
+        await _eventService.LogOrganizationUserEventsAsync(events.Select(e => (e.ou, e.e, _EventSystemUser, e.d)));
+    }
+
+    private async Task UpdateUsersAsync(Group group, HashSet<string> groupUsers,
+        Dictionary<string, Guid> existingUsersIdDict, HashSet<Guid> existingUsers = null)
+    {
+        var availableUsers = groupUsers.Intersect(existingUsersIdDict.Keys);
+        var users = new HashSet<Guid>(availableUsers.Select(u => existingUsersIdDict[u]));
+        if (existingUsers != null && existingUsers.Count == users.Count && users.SetEquals(existingUsers))
+        {
+            return;
+        }
+
+        await _groupRepository.UpdateUsersAsync(group.Id, users);
+    }
+
+    private async Task RemoveExistingExternalUsers(IEnumerable<string> removeUserExternalIds,
+            List<(OrganizationUserUserDetails ou, EventType e, DateTime? d)> events,
+            OrganizationUserImportData importUserData)
+    {
+        if (!removeUserExternalIds.Any())
+        {
+            return;
+        }
+
+        var existingUsersDict = importUserData.ExistingExternalUsers.ToDictionary(u => u.ExternalId);
+        var removeUsersSet = new HashSet<string>(removeUserExternalIds)
+            .Except(importUserData.NewUsersSet)
+            .Where(u => existingUsersDict.ContainsKey(u) && existingUsersDict[u].Type != OrganizationUserType.Owner)
+            .Select(u => existingUsersDict[u]);
+
+        await _organizationUserRepository.DeleteManyAsync(removeUsersSet.Select(u => u.Id));
+        events.AddRange(removeUsersSet.Select(u => (
+          u,
+          EventType.OrganizationUser_Removed,
+          (DateTime?)DateTime.UtcNow
+          ))
+        );
+    }
+
+    private async Task UpsertExistingUsers(Organization organization,
+            IEnumerable<ImportedOrganizationUser> newUsers,
+            OrganizationUserImportData importUserData
+            )
+    {
+        if (!newUsers.Any())
+        {
+            return;
+        }
+
+        // Marry existing users
+        var existingUsersEmailsDict = importUserData.ExistingUsers
+            .Where(u => string.IsNullOrWhiteSpace(u.ExternalId))
+            .ToDictionary(u => u.Email);
+        var newUsersEmailsDict = newUsers.ToDictionary(u => u.Email);
+        var newAndExistingUsersIntersection = existingUsersEmailsDict.Keys.Intersect(newUsersEmailsDict.Keys).ToList();
+        var organizationUsers = (await _organizationUserRepository.GetManyAsync(importUserData.ExistingUsers.Select(u => u.Id).ToList())).ToDictionary(u => u.Id);
+        var usersToUpsert = new List<OrganizationUser>();
+
+        foreach (var user in newAndExistingUsersIntersection)
+        {
+            var organizationUser = organizationUsers[existingUsersEmailsDict[user].Id];
```

Comment:
If the user doesn't exist, I think this will throw an exception. You want to use `TryGetValue()` instead